### DISCUSSION TITLE
LUYA CMS Multiple Website Domains and Sub-Domains

### DISCUSSION THREAD

**Discussion by nadar** [luyadev/luya#2138] [Questions & Problems]
Q:
how can i separate the frontend views in a different domain from the backend, as i need to have one cms backend and many pages with different domains?

A:
Hi, in order to find a solution for that questions i think you need to make your example more explicit.

1. Do you want to have different pages for different domains? Or are those different domains but the same cms content (like alias domains)?
2. Can you share the same database among those domains?
3. With `backend` i assume you mean the admin interface?

There is currently a huge PR pending from (@boehsermoe) regarding "multiple websites support" (which means, multiple domains which point to the same webserver), so if this is what you are looking for, its in the making https://github.com/luyadev/luya-module-cms/pull/274

Q:
1- Do you want to have different pages for different domains? Or are those different domains but the same cms content (like alias domains)?
answer: same cms content (like alias domains).

2- Can you share the same database among those domains?
answer: yes the same database

3- With backend i assume you mean the admin interface?
answer: yes luya admin interface

A:
So you would like to have:

* a.example.com => display content A
* b.example.com => display content A

Then you can just use https://luya.io/api/luya-web-Composition#$hostInfoMapping-detail

```php
'hostInfoMapping' => [
          'http://mydomain.com' => ['langShortCode' => 'en'],
          'http://meinedomain.de' => ['langShortCode' => 'en'],
 ],
```

This would be the behavior for alias domains. But when you like to separate the content from a domain, then you have to wait until the above mentioned PR is merged :slight_smile: 

* a.example.com => display content A
* b.example.com => display content B

Does this answer your questions? Otherwise you have to explain a little better what you like to do, maybe with concrete domains.

**Comment by hbugdoll** [comment]
> But when you like to separate the content from a domain, then you have to wait until the above mentioned PR is merged :slight_smile:
> 
> * a.example.com => display content A
> * b.example.com => display content B

This one will be the main use-case.

Since the docs are missing, I struggled several times handling multiple websites in LUYA – right at the setup:

- Installing LUYA at folder `luya/` at domain `example.com`, a default website is generated.
- Then create a subdomain e.g. `b.example.com` and forward it to `luya/public_html/` **?**
- In LUYA add a second website and set its 'host' property to "b.example.com" under _Settings_ > _Websites_.
- Which login page should then be used, https://example.com/admin or https://b.example.com/admin **?**

**Reply by nadar** [reply]
+ yes, create a domain alias (not fortwarding, but i assume this is what you mean)
+ which login: both should work

**Reply by hbugdoll** [reply]
Thx @nadar for your reply.
Ah, you mean "real" domain alias via a CNAME DNS record.

Using https://example.com/admin as login, everything is fine with the first website ✅.
But switching to the second website via `currentWebsiteToggler` in the CMS menu tree
then a significant problem occurred in page preview and live preview (`/cms-page-preview?itemId=...`) ⚡:

`Yii::$app->website->current` points to the "wrong" website, because `Website::getCurrent()` is host-driven, see also
https://github.com/luyadev/luya-module-cms/blob/8a3472acfc2620374fb8de6aed1f2736926cca2b/src/Website.php#L33

Otherwise using https://b.example.com/admin, everything is fine with the second website,
but the problem above occurred at the first website.

**Reply by nadar** [reply]
getCurrent() should resolve the current Website based on the current hostname. So what is the problem? Is there something i can help fixing it?

**Reply by hbugdoll** [reply]
> getCurrent() should resolve the current Website based on the current hostname.

But only if the host name has been explicitly entered under *Settings* > *Websites*
(the fall-back returns the default website).
We definitely need a hint to this in the new guide section.


> So what is the problem? Is there something i can help fixing it?

On published websites the host-based concept works as expected.
But in the admin area it's not possible to change `Website::getCurrent()`'s result via `currentWebsiteToggler`.

Let's take your example websites `a.example.com` (named A) and `b.example.com` (named B):
<img width="812" alt="Websites" src="https://github.com/user-attachments/assets/b7d19359-60c6-44ea-9e92-aac6b1a0b480">

Logged-in via https://a.example.com/admin `Website::getCurrent()` always returns website A:
<img width="824" alt="Website_a_A" src="https://github.com/user-attachments/assets/393946ca-efdf-4b74-9773-adb2e03db6a8">
<img width="822" alt="Website_a_B" src="https://github.com/user-attachments/assets/b2e9d451-cbc9-4d83-bca3-3a522e070169">

Logged-in via https://b.example.com/admin `Website::getCurrent()` always returns website B:
<img width="832" alt="Website_b_A" src="https://github.com/user-attachments/assets/025647fc-eced-4d4b-842d-bbf905cdf464">
<img width="834" alt="Website_b_B" src="https://github.com/user-attachments/assets/86fe77a6-683f-4459-97af-2afd7d59dcbc">

In my opinion `Website::getCurrent()` should reflect the state of `currentWebsiteToggler` in the admin area.

**Comment by hbugdoll** [comment]
I would also like to point out that changing from https://a.example.com/admin to https://b.example.com/admin (and vice versa) the admin must log in again with credentials.

----

Another problem I have noticed is if a website is not the default website, all menu item are "offline" in the preview and have to be injected, see https://github.com/luyadev/luya-module-cms/blob/8a3472acfc2620374fb8de6aed1f2736926cca2b/src/frontend/controllers/PreviewController.php#L45-L70

e.g.:
`is_default = false` &rarr; https://a.example.com/cms-page-preview?itemId=123&version=456 ⚠️ 
`is_default = true` &rarr; redirect to https://b.example.com/page-alias ✅
